File: leet-code/connect.js
/**
 * // Definition for a Node.
 * function Node(val, left, right, next) {
 *    this.val = val === undefined ? null : val;
 *    this.left = left === undefined ? null : left;
 *    this.right = right === undefined ? null : right;
 *    this.next = next === undefined ? null : next;
 * };
 */

/**
 * @param {Node} root
 * @return {Node}
 */
var connect = function (root) {
  if (root == null) {
    return null
  }
  let queue = []
  queue.push(root)
  while (queue.length > 0) {
    const levelSize = queue.length
    for (let i = 0; i < levelSize; i++) {
      const tempNode = queue[i]
      if (tempNode.left) {
        queue.push(tempNode.left)
      }
      if (tempNode.right) {
        queue.push(tempNode.right)
      }
      if (i === levelSize - 1) {
        tempNode.next = null
      } else {
        tempNode.next = queue[i + 1]
      }
    }
    queue = queue.slice(levelSize)
  }
  return root
}

File: leet-code/connect.go
package main

// Node ...
type Node struct {
	Val   int
	Left  *Node
	Right *Node
	Next  *Node
}

func connect(root *Node) *Node {
	if root == nil {
		return nil
	}
	queue := make([]*Node, 0)
	queue = append(queue, root)
	for len(queue) > 0 {
		levelSize := len(queue)
		for i := 0; i < levelSize; i++ {
			tempNode := queue[i]
			if tempNode.Left != nil {
				queue = append(queue, tempNode.Left)
			}
			if tempNode.Right != nil {
				queue = append(queue, tempNode.Right)
			}
			if i != levelSize-1 {
				tempNode.Next = queue[i+1]
			} else {
				tempNode.Next = nil
			}
		}
		queue = queue[levelSize:]
	}
	return root
}
